beacon-chain/p2p/peers/status.go

```diff
@@ -0,0 +1,311 @@
+		return fmt.Errorf("peer %v already exists", pid)
+	}
+
+	status := &peerStatus{
+		address:   address,
+		direction: direction,
+		peerState: PeerDisconnected,
+	}
+
+	p.status[pid] = status
+	return nil
+}
+
+// Address returns the multiaddress of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) Address(pid peer.ID) (ma.Multiaddr, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if status, ok := p.status[pid]; ok {
+		return status.address, nil
+	}
+	return nil, fmt.Errorf("peer %v unknown", pid)
+}
+
+// Direction returns the direction of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) Direction(pid peer.ID) (network.Direction, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if status, ok := p.status[pid]; ok {
+		return status.direction, nil
+	}
+	return network.DirUnknown, fmt.Errorf("peer %v unknown", pid)
+}
+
+// SetChainState sets the chain state of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) SetChainState(pid peer.ID, chainState *pb.Status) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if status, ok := p.status[pid]; ok {
+		status.chainState = chainState
+		return nil
+	}
+	return fmt.Errorf("peer %v unknown", pid)
+}
+
+// ChainState gets the chain state of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) ChainState(pid peer.ID) (*pb.Status, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if status, ok := p.status[pid]; ok {
+		if status.chainState == nil {
+			return nil, fmt.Errorf("peer %v has no known chain state", pid)
+		}
+		return status.chainState, nil
+	}
+	return nil, fmt.Errorf("peer %v unknown", pid)
+}
+
+// SetConnectionState sets the connection state of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) SetConnectionState(pid peer.ID, state PeerConnectionState) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if status, ok := p.status[pid]; ok {
+		status.peerState = state
+		status.chainStateLastUpdated = roughtime.Now()
+		return nil
+	}
+	return fmt.Errorf("peer %v unknown", pid)
+}
+
+// ConnectionState gets the connection state of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) ConnectionState(pid peer.ID) (PeerConnectionState, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if status, ok := p.status[pid]; ok {
+		return status.peerState, nil
+	}
+	return PeerDisconnected, fmt.Errorf("peer %v unknown", pid)
+}
+
+// ChainStateLastUpdated gets the last time the chain state of the given remote peer was updated.
+// This will error if the peer does not exist.
+func (p *Status) ChainStateLastUpdated(pid peer.ID) (time.Time, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if status, ok := p.status[pid]; ok {
+		return status.chainStateLastUpdated, nil
+	}
+	return time.Now(), fmt.Errorf("peer %v unknown", pid)
```

Comment:
Instead of using `time.Now()`, I recommend using `roughtime.Now()` which is our time wrapper using roughtime NTP servers. Most of the codebase uses roughtime.Now() and if there is a shift from the NTP then you might experience odd behavior.

Reply:
This return value really should be ignored, as it is an error condition, but there is no `nil` for  `time.Time` so I have to put something in there.

Reply:
Had a bit more of a think about this but short of returning a pointer to the `time.Time` and hence allowing a `nil` return value (which I'm not a fan of) I don't think we can do much about not returning a value.

Given we need to return a value `now()` is a better option than `0` or similar.  And as long as the callers follow the rule of checking the error first we'll be fine (no different from returning an int, for example).

That said, I've updated it to use `roughtime` so this doesn't jar visually when people are looking at the code.